### DISCUSSION TITLE
Queue::write_texture finished callback?

### DISCUSSION THREAD

**Discussion by MatrixDev** [gfx-rs/wgpu#3458] [General]
So I'm reading documentation and just can't understand how to know when texture upload has finished.

As I understood we need to use `on_submitted_work_done` to check when submitted operation has finished. But it is said in the `write_texture` documentation that:
```
This method is intended to have low performance costs. As such, the write is not immediately submitted,
and instead enqueued internally to happen at the start of the next submit() call.
```

Taking this to account `on_submitted_work_done` will not wait for `write_texture` but instead for some "random" previous `submit` call because no `submit` has actually happen.

Is there any other way to know when `write_texture` has finished instead of submitting buffers manually? I really like the convinience of this function and no "multiple of 256" limitation.

**Comment by kpreid** [comment]
Normally, you do not need to care exactly when `write_texture()` has finished; you enqueue draw/compute operations _after_ `write_texture()`, and you know that they will execute afterward and thus predictably be able to read the written texels.

If that's not sufficient, can you say more about your use case — exactly what operation needs to wait for the write?

**Comment by MatrixDev** [comment]
@kpreid, yes, I know that texture can be used right after creation even without calling `write_texture`.
Here are my use-cases:
1. "loading screen" until all initial textures are uploaded (before main scene is rendered)
2. Add some kind of animation after texture is uploaded to avoid pop-in (when main scene is already rendering)

**Reply by cwfitzgerald** [reply]
`write_texture` is defined such that the data will be available on the gpu for all commands that are submitted after the write_texture call returns to the caller.

**Reply by cwfitzgerald** [reply]
In both of your cases, as long as you have called write_texture, you can assume all future work will see it available.
